### ISSUE TITLE
error handler in StreamIO::write() assumes last_error is present

### ISSUE THREAD

**Issue by jjdunn** [php-amqplib/php-amqplib#1188]
**Version(s) affected**: 3.6.2
**PHP version**: 8.3.9
**RabbitMQ version**: 0.9.1

**Description**  
catch block in Wire/IO/StreamIO::write() assumes $this->last_error is set, which is untrue in this case

**How to reproduce**  
don't know - some error was thrown while trying to create a connection to a server which is generally available; but was offline in this situation

**Possible Solution**  
check for last_error before trying to do getCode()

/Wire/IO/StreamIO.php lines 264-273
```
try {
    // check stream and prevent from high CPU usage
    $result = 0;
    if ($this->select_write()) {
        $buffer = mb_substr($data, $written, self::BUFFER_SIZE, 'ASCII');
        $result = fwrite($this->sock, $buffer);
    }
    $this->throwOnError();
} catch (\ErrorException $e) {
    $code = $this->last_error->getCode();   <=== ERROR is HERE assuming $this->last_error is SET
```

**Additional context**  
partial stack trace from production system, with connection details REMOVED:
```
2024/08/03 02:03:41[error] [exception.Error] 

Error: Call to a member function getCode() on null in

C:\web\prod\app\extensions\PhpAmqpLib\Wire\IO\StreamIO.php:273

Stack trace:
#0
C:\web\prod\app\extensions\PhpAmqpLib\Connection\AbstractConnection.php(410):
PhpAmqpLib\Wire\IO\StreamIO->write('AMQP\x00\x00\t\x01')
#1
C:\web\prod\app\extensions\PhpAmqpLib\Connection\AbstractConnection.php(261):
PhpAmqpLib\Connection\AbstractConnection->write('AMQP\x00\x00\t\x01')
#2
C:\web\prod\app\extensions\PhpAmqpLib\Connection\AbstractConnection.php(235):
PhpAmqpLib\Connection\AbstractConnection->connect()
#3
C:\web\prod\app\extensions\PhpAmqpLib\Connection\AMQPStreamConnection.php(71):
PhpAmqpLib\Connection\AbstractConnection->__construct('REMOVED',
'REMOVED', 'REMOVED', false, 'AMQPLAIN',
Object(PhpAmqpLib\Wire\AMQPWriter), 'en_US',
Object(PhpAmqpLib\Wire\IO\StreamIO), 0, 3, 0.0, NULL)
#4
C:\web\prod\app\extensions\PhpAmqpLib\Connection\AMQPSSLConnection.php(35):
PhpAmqpLib\Connection\AMQPStreamConnection->__construct('REMOVED',
'5671', 'REMOVED', 'REMOVED', 'REMOVED',
false, 'AMQPLAIN', NULL, 'en_US', 3, 130, Resource id #69, false, 0,
0.0, NULL)
```

**Comment by ramunasd** [comment]
Can You debug where exactly exception was thrown?

**Comment by jjdunn** [comment]
> Can You debug where exactly exception was thrown?

The process which opens this AMQPS connection runs every 5 minutes, all day long, and generally works flawlessly.  I think the MQ server it connects to went offline in the middle of the connection process.  I can't recreate the problem - only point to the issue with catching/handling the exception.

see stack trace in the ticket description.  given that the error occurred in StreamIO::write() in the catch block (line 273), I have to assume it was the either in select_write() or the mb_substr() or fwrite() on lines 268-269. 

p.s. this code is running on Windows Server 2019 - should have mentioned that in the ticket.

**Comment by jjdunn** [comment]
I've been able to recreate this problem, some of the time, in my local/development environment.

screenshot of stack-trace in the catch block of StreamIO::write():
![stack-trace-StreamIO-write-catch-block](https://github.com/user-attachments/assets/217eabce-fbed-4806-ab64-c7d43d046bc6)

screenshot of StreamIO properties showing last_error is undefined/null:
![last-error-undefined](https://github.com/user-attachments/assets/2c217d6c-df97-4f5e-a226-7be565fe6785)

**Comment by ramunasd** [comment]
I see actual exception `$e` is of class ErrorException. Can You expand it and show error message together with code? Maybe previous exception too?

**Comment by ramunasd** [comment]
BTW release 3.7.0 has workaround for this issue - https://github.com/php-amqplib/php-amqplib/commit/3d318f610d3486b447400222a540e0d3458c523e#diff-7ebcbc9d208eace17950555344449b3e033eb6bfdabf2541fae2207ddfeb37f8R273

**Comment by jjdunn** [comment]
> I see actual exception `$e` is of class ErrorException. Can You expand it and show error message together with code number? Maybe previous exception too?

here's a screenshot of the ErrorException details.  previous is null; error is occurring in fwrite():
![StreamIO-write-exception-details](https://github.com/user-attachments/assets/c162c92d-1792-42c0-81f8-951d2e952a1e)

> BTW release 3.7.0 has workaround for this issue - 
thanks. That's exactly what I was hoping for.

Since fwrite() is a PHP function,, where does the code *expect* last_write to get populated?

**Comment by ramunasd** [comment]
Completely fixed in version 3.7.1
